herddb-core/src/main/java/herddb/core/TableManager.java

```diff
@@ -740,17 +740,26 @@ public void unload(long pageId) {
     private void flushNewPageForCheckpoint(DataPage page, DataPage spareDataPage) {
         final boolean flushed = flushNewPage(page, spareDataPage);
 
-        if (!flushed) {
+        if (flushed) {
+            /*
+             * Replace the page in memory with his immutable version (faster modification checks). We can
+             * replace the page even if we didn't flush it (i.e.:flushed by some other thread) because we are
+             * running during a checkpoint and no other page write could happen (thus our page copy is fully
+             * equivalent to really flushed one).
+             *
+             * We replace the page only if we have actually flushed it.. if it was flushed by another thread
+             * it was flushed due to unload request (no concurrent checkpoints) and then removed from pages
+             * (or the thread is going to remove it). We don't want to keep knowledge of a page not know
+             * anymore to page replacement policy (It can't be remove again)
+             *
+             * For similar reason we replace the page only if there actually is a page in the first place. If
+             * a concurrent thread flushed and removed the page we don't want to add it again.
+             */
+            pages.computeIfPresent(page.pageId, (i,p) -> page.toImmutable());
```

Comment:
` pages.computeIfPresent(page.pageId, (i,p) -> page.toImmutable());`

better form:
` pages.computeIfPresent(page.pageId, (i,p) -> p.toImmutable());`

in fact your first form creates an inner class (or something like that) that needs to "capture" the context,
the second form is a simple function with two arguments

Reply:
Practically speaking we could switch between the 2 implementations but can be p != page in some unfortunate cases. (But the should be fully equivalent).

I'll chage the  implementation and add an assertion to enforce it

Reply:
On second thought... adding an assertion require to have both p and page thus creating a "capturing" inner class